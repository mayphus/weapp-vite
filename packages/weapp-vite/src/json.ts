import type { App, Component, Page, Sitemap, Theme } from '@weapp-core/schematics'
import type { CompilerContext } from './context/index'

export type {
  App,
  Component,
  Page,
  Sitemap,
  Theme,
}

type ReturnSelf<T> = (config: T) => T

type ConfigFn<T> = T | ((ctx: CompilerContext) => T)

type DefineJsonFn<T> = ReturnSelf<ConfigFn<T>>

export const defineAppJson: DefineJsonFn<App> = (config) => {
  return config
}

export const definePageJson: DefineJsonFn<Page> = (config) => {
  return config
}

export const defineComponentJson: DefineJsonFn<Component> = (config) => {
  return config
}

export const defineSitemapJson: DefineJsonFn<Sitemap> = (config) => {
  return config
}

export const defineThemeJson: DefineJsonFn<Theme> = (config) => {
  return config
}
