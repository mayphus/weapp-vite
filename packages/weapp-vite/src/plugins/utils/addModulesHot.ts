import type { PluginContext } from 'rollup'

export function addModulesHot(entriesSet: Set<string>, pluginContext: PluginContext) {
  for (const entry of entriesSet) {
    const moduleInfo = pluginContext.getModuleInfo(entry)

    if (moduleInfo) {
      const stack = [moduleInfo.id] // 用栈模拟递归
      const visitedModules = new Set<string>()

      while (stack.length > 0) {
        const id = stack.pop()

        if (id && !visitedModules.has(id)) {
          visitedModules.add(id)

          const info = pluginContext.getModuleInfo(id)

          if (info) {
            pluginContext.addWatchFile(info.id)
            // 将子依赖加入栈
            stack.push(...info.importedIds)
          }
        }
      }
    }
  }
}
