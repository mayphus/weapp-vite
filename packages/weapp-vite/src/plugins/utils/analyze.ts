import type { App, Component, Page } from '@weapp-core/schematics'
import { get } from '@weapp-core/shared'

export function analyzeAppJson(json: App) {
  const entries: string[] = []
  const pages = json.pages ?? []
  const components = Object.values(json.usingComponents ?? {})
  const subPackages = (
    [...json.subPackages ?? [], ...json.subpackages ?? []]?.reduce<string[]>(
      (acc, cur) => {
        acc.push(...(cur.pages ?? []).map(x => `${cur.root}/${x}`))
        return acc
      },
      [],
    )) ?? []

  if (get(json, 'tabBar.custom')) {
    entries.push('custom-tab-bar/index')
  }
  // 全局工具栏
  // https://developers.weixin.qq.com/miniprogram/dev/framework/runtime/skyline/appbar.html
  if (get(json, 'appBar')) {
    entries.push('app-bar/index')
  }
  entries.push(
    ...pages,
    ...components,
    ...subPackages,
  )
  return entries
}

export function analyzeCommonJson(json: Page | Component) {
  const entries: string[] = []
  const components = Object.values(json.usingComponents ?? {})
  entries.push(...components)
  return entries
}
