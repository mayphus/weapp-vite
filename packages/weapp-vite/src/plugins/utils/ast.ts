import type { Node } from 'estree'
// import type { ProgramNode } from 'rollup'
import { walk } from 'estree-walker'

export function collectRequireTokens(ast: Node) {
  const requireModules: {
    start: number
    end: number
    value: string
    async?: boolean
    // leadingComment: string
  }[] = []
  walk(ast, {
    enter(node) {
      if (node.type === 'CallExpression') {
        // TODO 可能要被废弃
        // https://developers.weixin.qq.com/miniprogram/dev/reference/api/require.html#require-async-%E9%93%BE%E5%BC%8F%E8%B0%83%E7%94%A8
        if (node.callee.type === 'Identifier' && node.callee.name === 'require') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
            })
          }
        }
        else if (
          node.callee.type === 'MemberExpression'
          && node.callee.object.type === 'Identifier'
          && node.callee.object.name === 'require'
          && node.callee.property.type === 'Identifier'
          && node.callee.property.name === 'async') {
          if (node.arguments[0] && node.arguments[0].type === 'Literal' && typeof node.arguments[0].value === 'string') {
            requireModules.push({
              // @ts-ignore
              start: node.arguments[0].start,
              // @ts-ignore
              end: node.arguments[0].end,
              value: node.arguments[0].value,
              async: true,
            })
          }
        }
      }
    },
  })

  return {
    requireModules,
  }
}
