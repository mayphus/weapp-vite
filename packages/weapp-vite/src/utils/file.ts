import { configExtensions, jsExtensions, templateExtensions, vueExtensions } from '@/constants'
import fs from 'fs-extra'
import path from 'pathe'

export function isJsOrTs(name?: string) {
  if (typeof name === 'string') {
    return jsExtensions.some(x => name.endsWith(`.${x}`))
  }
  return false
}

export function isTemplateRequest(request: string) {
  return request.endsWith('.wxml') || request.endsWith('.html')
}

export function changeFileExtension(filePath: string, extension: string) {
  if (typeof filePath !== 'string') {
    throw new TypeError(`Expected \`filePath\` to be a string, got \`${typeof filePath}\`.`)
  }

  if (typeof extension !== 'string') {
    throw new TypeError(`Expected \`extension\` to be a string, got \`${typeof extension}\`.`)
  }

  if (filePath === '') {
    return ''
  }

  extension = extension ? (extension.startsWith('.') ? extension : `.${extension}`) : ''

  const basename = path.basename(filePath, path.extname(filePath))
  return path.join(path.dirname(filePath), basename + extension)
}

export async function findVueEntry(filepath: string) {
  for (const ext of vueExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsEntry(filepath: string) {
  for (const ext of jsExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findJsonEntry(filepath: string) {
  for (const ext of configExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export async function findTemplateEntry(filepath: string) {
  for (const ext of templateExtensions) {
    const p = changeFileExtension(filepath, ext)
    if (await fs.exists(p)) {
      return p
    }
  }
}

export function isTemplate(filepath: string) {
  return templateExtensions.some(ext => filepath.endsWith(`.${ext}`))
}

export function touch(filename: string) {
  const time = new Date()

  try {
    fs.utimesSync(filename, time, time)
  }
  catch {
    fs.closeSync(fs.openSync(filename, 'w'))
  }
}
