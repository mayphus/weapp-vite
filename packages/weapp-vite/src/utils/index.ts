export * from './file'
export * from './json'
export * from './projectConfig'
export * from './scan'
// export * from './wxml'
// export * from './wxs'
