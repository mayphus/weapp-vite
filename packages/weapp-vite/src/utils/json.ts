import { parse as parseJson, stringify } from 'comment-json'
import fs from 'fs-extra'
import logger from '../logger'

export function parseCommentJson(json: string) {
  return parseJson(json, undefined, true)
}

export async function readCommentJson(filepath: string) {
  try {
    return parseCommentJson(await fs.readFile(filepath, 'utf8'))
  }
  catch {
    logger.error(`残破的JSON文件: ${filepath}`)
  }
}

export function stringifyJson(value: unknown, replacer?: (
  (key: string, value: unknown) => unknown
) | Array<number | string> | null) {
  return stringify(value, replacer, 2)
}

export function resolveJson(value: unknown) {
  return stringifyJson(value)
}
