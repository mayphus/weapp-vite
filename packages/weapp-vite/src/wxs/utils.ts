import { addExtension, removeExtensionDeep } from '@weapp-core/shared'

export function normalizeWxsFilename(value: string) {
  return addExtension(removeExtensionDeep(value), '.wxs')
}
