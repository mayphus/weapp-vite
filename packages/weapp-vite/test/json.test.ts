import { stringifyJson } from '@/utils/json'

describe('json', () => {
  describe('stringifyJson', () => {
    it('resolve replacer case 0', () => {
      const res = stringifyJson({
        a: {
          b: '@/xx/yy',
        },
      })
      expect(res).toMatchSnapshot()
    })

    it('resolve replacer case 1', () => {
      const res = stringifyJson({
        a: {
          b: '@/xx/yy',
        },
      }, (_key, value) => {
        return value
      })
      expect(res).toMatchSnapshot()
    })
  })
})
