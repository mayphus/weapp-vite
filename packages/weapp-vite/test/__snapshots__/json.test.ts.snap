// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`json > stringifyJson > resolve replacer case 0 1`] = `
"{
  "a": {
    "b": "@/xx/yy"
  }
}"
`;

exports[`json > stringifyJson > resolve replacer case 1 1`] = `
"{
  "a": {
    "b": "@/xx/yy"
  }
}"
`;
